RFC Proposal: Bluetooth: Host: Pass pointer to server in L2CAP accept() callback

donatieng opened the pull request:
RFC Proposal: https://github.com/zephyrproject-rtos/zephyr/issues/58402.

This PR amends the existing L2CAP server accept() callback to pass the relevant `struct bt_l2cap_server*` as an argument. This allows writing modules which manage multiple servers/PSMs at runtime, as the shared `accept()` callback can now recover which specific server/PSM the connection request refers to.

Resolves: https://github.com/zephyrproject-rtos/zephyr/issues/58402

donatieng committed:
Bluetooth: Host: Pass pointer to server in L2CAP accept() callback

Add a pointer to the associated server structure in the L2CAP accept()
callback. This allows the callee to know which server an incoming L2CAP
connection is associated with.

Signed-off-by: Donatien Garnier <donatien.garnier@blecon.net>

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -453,14 +455,15 @@ struct bt_l2cap_server {
 	 *  @return -EPERM if encryption key size is too short.
 	 */
-	int (*accept)(struct bt_conn *conn, struct bt_l2cap_chan **chan);
+	int (*accept)(struct bt_conn *conn, struct bt_l2cap_server* server, struct bt_l2cap_chan **chan);
```
This will be a stable API change, and will need to follow the steps defined by https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#stable

donatieng committed:
Bluetooth: Host: Amend existing L2CAP accept callbacks

This patch amends the existing L2CAP accept callbacks to use the new
accept signature that includes a pointer to the L2CAP server structure.

Signed-off-by: Donatien Garnier <donatien.garnier@blecon.net>

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3634,7 +3634,8 @@ int bt_eatt_reconfigure(struct bt_conn *conn, uint16_t mtu)
 #endif /* CONFIG_BT_EATT */
 
-static int bt_eatt_accept(struct bt_conn *conn, struct bt_l2cap_chan **chan)
+static int bt_eatt_accept(struct bt_conn *conn, struct bt_l2cap_server *server,
+	struct bt_l2cap_chan **chan)
```
General comment: The indentation should have been
```suggestion
static int bt_eatt_accept(struct bt_conn *conn, struct bt_l2cap_server *server,
			  struct bt_l2cap_chan **chan)
```

This applies to multiple places in this PR

donatieng replied:
Run clang-format over the PR, should be now fixed

alwa-nordic reviewed (changes requested):
@donatieng: This PR needs a cleanup. Will you remove unintentional changes and run clang-format on the modified lines? There are some test/build failures returned from the CI as well. Will you look into that?

donatieng commented:
The PR should now be cleaned up; the missing CI failure seems unrelated to the PR itself (looks like this different PR gets the same failing test: https://github.com/zephyrproject-rtos/zephyr/pull/61166).

carlescufi commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -427,6 +427,8 @@ struct bt_l2cap_chan_ops {
  */
 #define BT_L2CAP_SDU_CHAN_SEND_RESERVE (BT_L2CAP_SDU_BUF_SIZE(0))
 
+struct bt_l2cap_server;
```
@donatieng is this really needed given that you use the struct inside the actual declaration? Can you try without?

alwa-nordic reviewed (changes requested):
Looks good, but [please remove the extra forward declaration]. It's not needed.

[please remove the extra forward declaration]: https://github.com/zephyrproject-rtos/zephyr/pull/60536/files#r1297070300

donatieng commented:
I've removed the forward declaration. It seems that the BabbleSIM tests are quite broken though.

Thalley commented:
> I've removed the forward declaration. It seems that the BabbleSIM tests are quite broken though.

Fixed by https://github.com/zephyrproject-rtos/zephyr/pull/61733

Thalley commented:
@donatieng Do you plan on updating this PR with `user_data` rather the `l2cap_server`?

donatieng commented:
@Thalley Please see https://github.com/zephyrproject-rtos/zephyr/issues/58402#issuecomment-1693093426

Thalley commented:
> @Thalley Please see [#58402 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/58402#issuecomment-1693093426)

> So in the interest of maintaining consistency with the L2CAP API in order to minimize changes I'll leave the PR as is now, however I will still mention the user_data option which I agree looks preferable, but probably would lead to more breaking changes!

I am not sure I agree with this statement

Whether you add a `void *user_data` or a `struct bt_l2cap_server *server` in the callback, will be equally large change and equally breaking. The only difference is that the former is following the API Design guidelines and is more flexible, and the latter is not. 

w.r.t. to consistency in the L2CAP API, you do have a point as the other callbacks provide a L2CAP specific pointer (`struct bt_l2cap_chan *chan`), but since we don't have existing APIs that provide a `struct bt_l2cap_server *server` I am unsure about the level on consistency we potentially maintain. 

I'd like to get input from @jhedberg and @alwa-nordic regarding providing a `struct bt_l2cap_server *server` versus a `void *user_data`. Either way will work though.

carlescufi reviewed (changes requested):
Please add a note to the migration guide

carlescufi commented:
> Please add a note to the migration guide

Will be done in a follow-up PR

donatieng commented:
Thanks @carlescufi, I'll prepare the migration guide PR
